src/ass/kfx.py

```diff
@@ -0,0 +1,237 @@
+from collections.abc import Sequence, Mapping
+
+from datetime import timedelta
+from pyass import Color
+from pyass import Alignment
+import pyass.tag
+
+from .consts import *
+from ..models.karaoke import *
+
+
+def get_line_format(line: KLine) -> pyass.EventFormat:
+    return (
+        pyass.EventFormat.COMMENT
+        if line.isAlone and line.isEN
+        else pyass.EventFormat.DIALOGUE
+    )
+
+
+def get_romaji_pos_tag(line: KLine) -> pyass.Tag:
+    if line.isSecondary:
+        return SECONDARY_ROMAJI_POS_TAG
+    elif line.isAlone:
+        return NO_EN_ROMAJI_POS_TAG
+    else:
+        return ROMAJI_POS_TAG
+
+
+def get_en_pos_tag(line: KLine) -> pyass.Tag:
+    return SECONDARY_EN_POS_TAG if line.isSecondary else EN_POS_TAG
+
+
+def get_char_transform_tags(
+    kChar: KChar,
+    switchDuration: timedelta,
+    transitionDuration: timedelta,
+    startTag: Sequence[pyass.Tag] = [pyass.AlphaTag(0xFF)],
+    enterTag: Sequence[pyass.Tag] = [pyass.AlphaTag(0x00)],
+    exitTag: Sequence[pyass.Tag] = [pyass.AlphaTag(0xFF)],
+) -> Sequence[pyass.Tag]:
+    return [
+        *startTag,
+        get_enter_transition_tag(kChar, switchDuration, transitionDuration, enterTag),
+        get_exit_transition_tag(kChar, switchDuration, transitionDuration, exitTag),
+    ]
+
+
+def get_enter_transition_tag(
+    kChar: KChar,
+    switchDuration: timedelta,
+    transitionDuration: timedelta,
+    resultTag: Sequence[pyass.Tag] = [pyass.AlphaTag(0x00)],
+) -> pyass.Tag:
+    return pyass.TransformTag(
+        start=kChar.fadeOffset, end=switchDuration + kChar.fadeOffset, to=resultTag
+    )
+
+
+def get_exit_transition_tag(
+    kChar: KChar,
+    switchDuration: timedelta,
+    transitionDuration: timedelta,
+    resultTag: Sequence[pyass.Tag] = [pyass.AlphaTag(0xFF)],
+) -> pyass.Tag:
+    return pyass.TransformTag(
+        start=kChar.line.duration
+        - transitionDuration
+        + switchDuration
+        + kChar.fadeOffset,
+        end=kChar.line.duration
+        - transitionDuration
+        + 2 * switchDuration
+        + kChar.fadeOffset,
+        to=resultTag,
+    )
+
+
+def get_char_actor_tag(
+    kChar: KChar,
+    actorToStyle: Mapping[str, Sequence[pyass.Tag]],
+    switchDuration: timedelta,
+) -> Sequence[pyass.Tag]:
+    if len(kChar.line.actorSwitches) == 0:
+        return []
+    return [*(actorToStyle[kChar.line.startActor])] + [
```

Comment:
Just `list(actorToStyle[kChar.line.startActor])` instead?

Reply:
Oh yeah that works whoops